src/AuthRoles.js

```diff
@@ -0,0 +1,322 @@
+const RestQuery = require('./RestQuery');
+import  _ from "lodash";
+import Auth from "./Auth";
+
+// operation result for role
+const OppResult = {
+  rejected: 0,      // role rejected (no path to role was valid)
+  accepted: 1,      // role accepted (at least one path to role was valid)
+  inconclusive: 2,  // circular
+  processing: 3     // role is being validated (this prevents circular roles)
+}
+
+// add only uniquely to array
+const addIfNeed = function(array, string){
+  if(_.indexOf(array, string) === -1){
+    array.push(string)
+  }
+}
+
+export function AuthRoles(auth: Auth, master, isMaster = false) {
+  this.auth = auth
+  this.userId = auth.user ? auth.user.id : undefined
+  this.master = master
+  this.isMaster = isMaster;
```

Comment:
What is the difference between those two? Why do we need both?

Reply:
Actually i dont think we do. I am working on another branch in which I removed all these unneeded and duplicate checks.